File: scripts/lto/commands/resume.py
"""lto resume — 跨 session 断点热启动。"""

from __future__ import annotations

import argparse, sys
from pathlib import Path

from .. import state as st
from .. import git_state as gs
from .. import artifacts as af


def run(args: argparse.Namespace) -> int:
    repo = args.repo.resolve()

    # Resolve run_id
    current_file = repo / ".lto" / "current"
    if args.run_id:
        run_id = st.validate_run_id(args.run_id)
    elif current_file.exists():
        run_id = current_file.read_text(encoding="utf-8").strip()
        if run_id:
            run_id = st.validate_run_id(run_id)
        else:
            print("LTO: no active run (empty .lto/current)", file=sys.stderr)
            return 1
    else:
        print("LTO: no active run (missing .lto/current)", file=sys.stderr)
        return 1

    state_path = repo / ".lto" / run_id / "state.json"
    state = st.load_state(state_path)
    if state is None:
        print(f"LTO: no state.json found for {run_id}", file=sys.stderr)
        return 1
    is_closed = state.get("current_phase") == "closed"

    # Validate workspace
    ws = state.get("workspace", {})
    recorded_head = ws.get("head", "unknown")
    actual_head = gs.git_head(repo)
    actual_branch = gs.git_branch(repo)
    drift = gs.head_drift(repo, recorded_head)
    dirty = gs.git_dirty(repo)

    # Determine resume status
    revalidate_tasks = []
    warnings = []

    if drift == "none":
        pass
    elif drift == "forward":
        # HEAD advanced — check if related files changed
        file_drift = gs.task_file_drift(repo, recorded_head, actual_head, state)
        if file_drift["changed_paths"]:
            revalidate_tasks = [t["id"] for t in state.get("tasks", []) if t["status"] in ("in_progress", "done")]
            sample = ", ".join(file_drift["changed_paths"][:6])
            warnings.append(
                f"HEAD advanced ({recorded_head[:8]}→{actual_head[:8]}), "
                f"related files changed: {sample}"
            )
        elif file_drift["missing_touched_files"]:
            warnings.append(
                f"HEAD advanced ({recorded_head[:8]}→{actual_head[:8]}), "
                "no task touched_files recorded; file drift precision unavailable"
            )
        else:
            warnings.append(f"HEAD advanced ({recorded_head[:8]}→{actual_head[:8]}), no related file changes")
    elif drift == "rewrite":
        revalidate_tasks = [t["id"] for t in state.get("tasks", []) if t["status"] in ("in_progress", "done")]
        warnings.append(f"HEAD rewritten ({recorded_head[:8]} not ancestor of {actual_head[:8]})")
    elif drift == "unreachable":
        warnings.append(f"recorded HEAD {recorded_head[:8]} unreachable")
        revalidate_tasks = [t["id"] for t in state.get("tasks", []) if t["status"] != "pending"]

    if dirty:
        warnings.append("worktree has uncommitted changes outside .lto")

    if is_closed:
        if revalidate_tasks:
            warnings.append(
                "run is closed; resume is read-only and will not reopen tasks "
                "or update recorded HEAD"
            )
            revalidate_tasks = []
        _print_capsule(repo, run_id, state, warnings, revalidate_tasks)
        return 0

    # Update state
    ws["head"] = actual_head
    ws["branch"] = actual_branch
    ws["dirty_fingerprint"] = "dirty" if dirty else "clean"
    state["workspace"] = ws

    # Mark revalidate
    for task_id in revalidate_tasks:
        try:
            st.update_task(state, task_id, status="pending", blockers=[{
                "reason": "requires_revalidate",
                "recorded_head": recorded_head,
                "current_head": actual_head,
            }])
        except KeyError:
            pass

    # Save updated state
    st.save_state(state_path, state)
    st.sync_run_state_md(repo / ".lto" / run_id / "run-state.md", state)
    af.load_manifest(repo, run_id, state=state, persist_synthesized=True)

    # Print context capsule
    _print_capsule(repo, run_id, state, warnings, revalidate_tasks)

    return 0 if not revalidate_tasks else 2


def _print_capsule(repo: Path, run_id: str, state: dict, warnings: list[str], revalidate: list[str]) -> None:
    phase = state.get("current_phase", "unknown")
    ws = state.get("workspace", {})
    head = ws.get("head", "unknown")

    tasks = state.get("tasks", [])
    task_summary = ", ".join(
        f"{t['id']}:{t['status']}" for t in tasks[-5:]
    ) if tasks else "none"

    last_failure = state.get("last_failure", "")
    next_action = state.get("next_action", "none")
    blocked = state.get("blocked_by", "none")
    unresolved = state.get("gates", {}).get("unresolved_blocks", [])

    print("=== LTO ACTIVE SESSION ===")
    print(f"Run ID: {state.get('run_id', '?')}")
    print(f"Goal: {state.get('goal', '?')}")
    print(f"Phase: {phase}")
    print(f"Head: {head[:12]} ({ws.get('dirty_fingerprint', '?')})")
    print(f"Tasks: {task_summary}")
    if last_failure:
        print(f"Last Failure: {last_failure[:120]}")
    print(f"Next: {next_action[:120]}")
    if blocked and blocked != "none":
        print(f"Blocked: {blocked}")
    if unresolved:
        print(f"Unresolved Blocks: {len(unresolved)}")

    entries = af.recent(repo, run_id, limit=6)
    if entries:
        print("Recent Artifacts:")
        for entry in entries:
            summary = entry.get("summary") or ""
            suffix = f" — {summary[:80]}" if summary else ""
            source = " (synthesized)" if entry.get("source") == "synthesized" else ""
            print(f"  - {entry.get('kind')}: {entry.get('relative_path')}{suffix}{source}")

    if warnings:
        print(f"\nWarnings:")
        for w in warnings:
            print(f"  ⚠ {w}")

    if revalidate:
        print(f"\n⚠ {len(revalidate)} tasks require revalidation: {', '.join(revalidate[:5])}")

    # 感知面：跨 session 回来的第一个命令必须重注入 affordance 事实——
    # SKILL.md 早出 context 窗口，插件不在这里可见就是死数据。零推荐，匹配归 host。
    try:
        from .. import plugins as plg
        aff = plg.affordance_facts(repo, run_id)
        if aff["available"]:
            mounted = set(aff["mounted"])
            names = ", ".join(
                f"{p['id']}{' (mounted)' if p['id'] in mounted else ''}"
                for p in aff["available"]
            )
            print(f"Plugins: {names}")
            print("  → 任务形态先验见 references/workflow-playbook.md；"
                  "细节 `lto plugin list`；挂载 `lto plugin mount <dir> --run-id <id>`")
    except Exception:
        pass  # 感知层绝不弄崩 capsule

    # 长 gap 提醒：人类侧 goal drift。距上次活动 >24h 时，提示给用户跑 recap。
    # capsule 本身是给 AI 的（git head/dirty），但隔了很久回来的人可能忘了在做什么。
    gap_hours = _max_session_gap_hours(state)
    if gap_hours >= 24:
        d = int(gap_hours / 24)
        print(f"\n⏳ 距上次活动约 {int(gap_hours)} 小时（{d} 天）。"
              f"建议先给用户跑 `lto recap`——隔这么久，人可能忘了在做什么、为什么。")

    print("===========================")


def _max_session_gap_hours(state: dict) -> float:
    """phase_transitions 相邻时间戳的最大间隔（小时）——近似 session gap。"""
    from datetime import datetime
    transitions = state.get("phase_transitions", [])
    prev = None
    max_gap = 0.0
    for tr in transitions:
        at = tr.get("at", "")
        if not at:
            continue
        try:
            dt = datetime.fromisoformat(at)
        except ValueError:
            continue
        if prev is not None:
            max_gap = max(max_gap, (dt - prev).total_seconds() / 3600)
        prev = dt
    return max_gap


def add_parser(subparsers) -> None:
    p = subparsers.add_parser("resume", help="resume from last checkpoint")
    p.add_argument("--run-id")
    p.set_defaults(func=run)

File: src/commands/resume.rs
use crate::commands::util;
use serde_json::Value;
use std::collections::BTreeSet;
use std::path::Path;

#[derive(Debug, Clone, Default)]
pub struct ResumeOptions {
    pub run_id: Option<String>,
}

pub fn cmd_resume(repo: &Path, options: ResumeOptions) -> anyhow::Result<()> {
    let ctx = util::load_run(repo, options.run_id.as_deref())?;
    let actual = util::git_status(repo);
    let recorded_head = if ctx.state.workspace.head.trim().is_empty() {
        "unknown"
    } else {
        ctx.state.workspace.head.as_str()
    };
    let drift = util::head_drift(repo, recorded_head, &actual.head);
    let mut warnings = Vec::new();
    let mut revalidate = Vec::new();

    match drift {
        "none" => {}
        "forward" => {
            let changed = util::git_changed_paths(repo, recorded_head, &actual.head);
            let touched = touched_files(&ctx.state.tasks);
            let related = changed
                .iter()
                .filter(|path| touched.contains(path.as_str()))
                .take(6)
                .cloned()
                .collect::<Vec<_>>();
            if !related.is_empty() {
                revalidate = revalidatable_tasks(&ctx.state.tasks);
                warnings.push(format!(
                    "HEAD advanced ({}→{}), related files changed: {}",
                    short(recorded_head),
                    short(&actual.head),
                    related.join(", ")
                ));
            } else if touched.is_empty() {
                warnings.push(format!(
                    "HEAD advanced ({}→{}), no task touched_files recorded; file drift precision unavailable",
                    short(recorded_head),
                    short(&actual.head)
                ));
            } else {
                warnings.push(format!(
                    "HEAD advanced ({}→{}), no related file changes",
                    short(recorded_head),
                    short(&actual.head)
                ));
            }
        }
        "rewrite" => {
            revalidate = revalidatable_tasks(&ctx.state.tasks);
            warnings.push(format!(
                "HEAD rewritten ({} not ancestor of {})",
                short(recorded_head),
                short(&actual.head)
            ));
        }
        "unreachable" => {
            revalidate = non_pending_tasks(&ctx.state.tasks);
            warnings.push(format!(
                "recorded HEAD {} unreachable",
                short(recorded_head)
            ));
        }
        _ => {}
    }
    if actual.dirty {
        warnings.push("worktree has uncommitted changes outside .lto".to_string());
    }
    if ctx.state.current_phase == "closed" && !revalidate.is_empty() {
        warnings.push(
            "run is closed; resume is read-only and will not reopen tasks or update recorded HEAD"
                .to_string(),
        );
        revalidate.clear();
    }

    print_capsule(repo, &ctx, &warnings, &revalidate);
    if revalidate.is_empty() {
        Ok(())
    } else {
        anyhow::bail!(
            "{} tasks require revalidation: {}",
            revalidate.len(),
            revalidate.join(", ")
        )
    }
}

fn print_capsule(repo: &Path, ctx: &util::RunContext, warnings: &[String], revalidate: &[String]) {
    let state = &ctx.state;
    let head = if state.workspace.head.trim().is_empty() {
        "unknown"
    } else {
        state.workspace.head.as_str()
    };
    let task_summary = task_summary(&state.tasks);
    let last_failure = value_as_line(&state.last_failure, 120);
    let next_action = value_as_line(&state.next_action, 120);
    let blocked = state.blocked_by.as_str().unwrap_or("none");
    let unresolved = state
        .gates
        .get("unresolved_blocks")
        .and_then(Value::as_array)
        .map(Vec::len)
        .unwrap_or_default();

    println!("=== LTO ACTIVE SESSION ===");
    println!("Run ID: {}", state.run_id);
    println!("Goal: {}", fallback(&state.goal, "?"));
    println!("Phase: {}", fallback(&state.current_phase, "unknown"));
    println!(
        "Head: {} ({})",
        short(head),
        fallback(&state.workspace.dirty_fingerprint, "?")
    );
    println!("Tasks: {task_summary}");
    if !last_failure.is_empty() {
        println!("Last Failure: {last_failure}");
    }
    println!("Next: {next_action}");
    if !blocked.is_empty() && blocked != "none" {
        println!("Blocked: {blocked}");
    }
    if unresolved > 0 {
        println!("Unresolved Blocks: {unresolved}");
    }

    let artifacts = util::latest_artifacts(repo, &ctx.run_id, 6);
    if !artifacts.is_empty() {
        println!("Recent Artifacts:");
        for entry in artifacts {
            let kind = entry.get("kind").and_then(Value::as_str).unwrap_or("other");
            let path = entry
                .get("relative_path")
                .and_then(Value::as_str)
                .or_else(|| entry.get("run_relative_path").and_then(Value::as_str))
                .unwrap_or("?");
            let summary = entry.get("summary").and_then(Value::as_str).unwrap_or("");
            let source = if entry.get("source").and_then(Value::as_str) == Some("synthesized") {
                " (synthesized)"
            } else {
                ""
            };
            if summary.is_empty() {
                println!("  - {kind}: {path}{source}");
            } else {
                println!("  - {kind}: {path} — {}{source}", truncate(summary, 80));
            }
        }
    }

    if !warnings.is_empty() {
        println!();
        println!("Warnings:");
        for warning in warnings {
            println!("  ⚠ {warning}");
        }
    }
    if !revalidate.is_empty() {
        println!();
        println!(
            "⚠ {} tasks require revalidation: {}",
            revalidate.len(),
            revalidate
                .iter()
                .take(5)
                .cloned()
                .collect::<Vec<_>>()
                .join(", ")
        );
    }

    let gap = util::max_session_gap_hours(state);
    if gap >= 24.0 {
        println!();
        println!(
            "⏳ 距上次活动约 {} 小时（{} 天）。建议先给用户跑 `lto recap`——隔这么久，人可能忘了在做什么、为什么。",
            gap as u64,
            gap as u64 / 24
        );
    }
    println!("===========================");
}

fn task_summary(tasks: &Value) -> String {
    let tasks = util::json_array(tasks);
    if tasks.is_empty() {
        return "none".to_string();
    }
    tasks
        .iter()
        .rev()
        .take(5)
        .collect::<Vec<_>>()
        .into_iter()
        .rev()
        .map(|task| {
            let id = task.get("id").and_then(Value::as_str).unwrap_or("?");
            let status = task
                .get("status")
                .and_then(Value::as_str)
                .unwrap_or("unknown");
            format!("{id}:{status}")
        })
        .collect::<Vec<_>>()
        .join(", ")
}

fn touched_files(tasks: &Value) -> BTreeSet<String> {
    util::json_array(tasks)
        .iter()
        .flat_map(|task| {
            task.get("touched_files")
                .and_then(Value::as_array)
                .cloned()
                .unwrap_or_default()
        })
        .filter_map(|value| value.as_str().map(str::to_string))
        .collect()
}

fn revalidatable_tasks(tasks: &Value) -> Vec<String> {
    util::json_array(tasks)
        .iter()
        .filter(|task| {
            matches!(
                task.get("status").and_then(Value::as_str),
                Some("in_progress" | "done")
            )
        })
        .filter_map(|task| task.get("id").and_then(Value::as_str).map(str::to_string))
        .collect()
}

fn non_pending_tasks(tasks: &Value) -> Vec<String> {
    util::json_array(tasks)
        .iter()
        .filter(|task| task.get("status").and_then(Value::as_str) != Some("pending"))
        .filter_map(|task| task.get("id").and_then(Value::as_str).map(str::to_string))
        .collect()
}

fn value_as_line(value: &Value, limit: usize) -> String {
    let raw = value.as_str().map(str::to_string).unwrap_or_else(|| {
        if value.is_null() {
            String::new()
        } else {
            value.to_string()
        }
    });
    truncate(&util::single_line(&raw), limit)
}

fn fallback<'a>(value: &'a str, fallback: &'a str) -> &'a str {
    if value.trim().is_empty() {
        fallback
    } else {
        value
    }
}

fn short(value: &str) -> String {
    value.chars().take(12).collect()
}

fn truncate(value: &str, limit: usize) -> String {
    value.chars().take(limit).collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::state::{self, LtoState, WorkspaceSnapshot};
    use serde_json::json;
    use std::fs;
    use std::process::Command;

    struct GitHarness {
        _tmp: tempfile::TempDir,
        repo: std::path::PathBuf,
    }

    impl GitHarness {
        fn new() -> Self {
            let tmp = tempfile::tempdir().unwrap();
            let repo = tmp.path().join("repo");
            fs::create_dir_all(&repo).unwrap();
            git(&repo, &["init"]);
            git(&repo, &["config", "user.email", "lto@example.test"]);
            git(&repo, &["config", "user.name", "LTO Test"]);
            Self { _tmp: tmp, repo }
        }

        fn write_commit(&self, file: &str, text: &str, msg: &str) -> String {
            fs::write(self.repo.join(file), text).unwrap();
            git(&self.repo, &["add", file]);
            git(&self.repo, &["commit", "-m", msg]);
            head(&self.repo)
        }

        fn write_state(&self, recorded_head: &str, tasks: serde_json::Value) {
            let run_dir = self.repo.join(".lto").join("r1");
            fs::create_dir_all(&run_dir).unwrap();
            fs::write(self.repo.join(".lto").join("current"), "r1\n").unwrap();
            let state = LtoState {
                run_id: "r1".to_string(),
                goal: "resume drift".to_string(),
                current_phase: "implementation".to_string(),
                workspace: WorkspaceSnapshot {
                    head: recorded_head.to_string(),
                    dirty_fingerprint: "clean".to_string(),
                    ..WorkspaceSnapshot::default()
                },
                tasks,
                next_action: json!("continue"),
                blocked_by: json!("none"),
                ..LtoState::default()
            };
            state::save_state(run_dir.join("state.json"), &state).unwrap();
        }
    }

    fn git(repo: &Path, args: &[&str]) {
        let out = Command::new("git")
            .args(args)
            .current_dir(repo)
            .output()
            .unwrap();
        assert!(
            out.status.success(),
            "git {:?}: {}",
            args,
            String::from_utf8_lossy(&out.stderr)
        );
    }

    fn head(repo: &Path) -> String {
        String::from_utf8(
            Command::new("git")
                .args(["rev-parse", "HEAD"])
                .current_dir(repo)
                .output()
                .unwrap()
                .stdout,
        )
        .unwrap()
        .trim()
        .to_string()
    }

    #[test]
    fn cmd_resume_main_path_has_no_revalidation_when_head_matches() {
        let h = GitHarness::new();
        let first = h.write_commit("a.txt", "a\n", "first");
        h.write_state(
            &first,
            json!([{"id": "T1", "status": "done", "touched_files": ["a.txt"]}]),
        );
        cmd_resume(
            &h.repo,
            ResumeOptions {
                run_id: Some("r1".into()),
            },
        )
        .unwrap();
    }

    #[test]
    fn cmd_resume_forward_drift_revalidates_related_done_tasks() {
        let h = GitHarness::new();
        let first = h.write_commit("a.txt", "a\n", "first");
        h.write_state(
            &first,
            json!([{"id": "T1", "status": "done", "touched_files": ["a.txt"]}]),
        );
        h.write_commit("a.txt", "changed\n", "change touched");
        let err = cmd_resume(
            &h.repo,
            ResumeOptions {
                run_id: Some("r1".into()),
            },
        )
        .unwrap_err();
        assert!(err.to_string().contains("T1"));
    }

    #[test]
    fn cmd_resume_rewrite_drift_revalidates_done_tasks() {
        let h = GitHarness::new();
        let first = h.write_commit("base.txt", "base\n", "base");
        let main_branch = String::from_utf8(
            Command::new("git")
                .args(["branch", "--show-current"])
                .current_dir(&h.repo)
                .output()
                .unwrap()
                .stdout,
        )
        .unwrap()
        .trim()
        .to_string();
        git(&h.repo, &["checkout", "-b", "side"]);
        let side = h.write_commit("side.txt", "side\n", "side");
        git(&h.repo, &["checkout", &main_branch]);
        assert_eq!(head(&h.repo), first);
        h.write_state(
            &side,
            json!([{"id": "T1", "status": "done", "touched_files": ["side.txt"]}]),
        );
        h.write_commit("main.txt", "main\n", "main");
        let err = cmd_resume(
            &h.repo,
            ResumeOptions {
                run_id: Some("r1".into()),
            },
        )
        .unwrap_err();
        assert!(err.to_string().contains("T1"));
    }

    #[test]
    fn cmd_resume_unreachable_recorded_head_revalidates_non_pending_tasks() {
        let h = GitHarness::new();
        h.write_commit("a.txt", "a\n", "first");
        h.write_state(
            "1111111111111111111111111111111111111111",
            json!([
                {"id": "T1", "status": "done"},
                {"id": "T2", "status": "pending"}
            ]),
        );
        let err = cmd_resume(
            &h.repo,
            ResumeOptions {
                run_id: Some("r1".into()),
            },
        )
        .unwrap_err();
        assert!(err.to_string().contains("T1"));
        assert!(!err.to_string().contains("T2"));
    }
}
